# Bioinformatics Code Hub
A curated collection of code for bioinformatic analyses for anyone to learn and copy for their analyses.

If you can't find an analysis you want to perform or if you want an analysis expanded, open an issue and the contributors will make a template for you.

The following analyses are planned:
1. Genomics
   - Copy number variation analysis
   - GWAS (Genome-wide association studies)
2. Transcriptomics
   - FASTQ processing
      - Quality control (FastQC)
      - Trimming (Trimmomatic)
      - Genome alignment
        - STAR
        - Kallisto
      - Transcriptome alignment (HTSeq)
   - Bulk RNA-Seq
     - Differential expression
        - Volcano plots
     - Pathway analysis
   - Single-Cell RNA-Seq
     - Cell Annotations
     - Compare cells from two different datasets
     - ✅Transcription factor analysis
     - ✅Psuedotime 
     - ✅Ligand-receptor interactions
   - Spatial Transcriptomics
     - ✅Visualize clusters on spatial coordinates
3. Proteomics
   - Mass spectrometry 
   - Protein-protein interactions
   - Post-translational modifications
4. Metabolomics
   - Nuclear Magnetic Resonance (NMR) spectroscopy
5. Epigenetics
   - DNA methylation
   - Chromatin accessibility analysis (ATAC-seq)
6. Metagenomics
   - 16S rRNA sequencing
   - Microbiome diversity
7. Computer Vision
   - Microscopy segmentation
   - MRI
8. Molecular Dynamics
   - MDVerse

Datasets used:
 - Single-cell RNA-Seq:
    - https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE109816
 - Spatial transcriptomics:
    - https://data.mendeley.com/datasets/mbvhhf8m62/2

 
Contributions are welcome ❤️
