{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PyDESeq2 Analysis\n",
    "\n",
    "This is the Python version of the popular DESeq2 package in R for differential gene expression analysis.\n",
    "DESeq2 is widely used for bulk RNA-seq count data, offering robust statistical methods to identify genes that are differentially expressed between experimental conditions.\n",
    "The goal is to take raw count data, normalize it, and extract insights from significant changes in gene expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load the bulk RNA-Seq data\n",
    "# Here we have the raw counts, which is the exact input we want for DESeq2.\n",
    "# Raw counts are generated by tools such as featureCounts or HTSeq-count.\n",
    "counts_path <- file.path(\"..\",\n",
    "                         \"data\",\n",
    "                         \"GSE164073_raw_counts_GRCh38.p13_NCBI.tsv.gz\")\n",
    "counts <- read.delim((counts_path))\n",
    "\n",
    "# Observe the data\n",
    "head(counts)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
