{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PyDESeq2 Analysis\n",
    "\n",
    "This is the Python version of the popular DESeq2 package in R for differential gene expression analysis.\n",
    "DESeq2 is widely used for bulk RNA-seq count data, offering robust statistical methods to identify genes that are differentially expressed between experimental conditions.\n",
    "The goal is to take raw count data, normalize it, and extract insights from significant changes in gene expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GSM4996084</th>\n",
       "      <th>GSM4996085</th>\n",
       "      <th>GSM4996086</th>\n",
       "      <th>GSM4996087</th>\n",
       "      <th>GSM4996088</th>\n",
       "      <th>GSM4996089</th>\n",
       "      <th>GSM4996090</th>\n",
       "      <th>GSM4996091</th>\n",
       "      <th>GSM4996092</th>\n",
       "      <th>GSM4996093</th>\n",
       "      <th>GSM4996094</th>\n",
       "      <th>GSM4996095</th>\n",
       "      <th>GSM4996096</th>\n",
       "      <th>GSM4996097</th>\n",
       "      <th>GSM4996098</th>\n",
       "      <th>GSM4996099</th>\n",
       "      <th>GSM4996100</th>\n",
       "      <th>GSM4996101</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100287102</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>653635</td>\n",
       "      <td>244</td>\n",
       "      <td>236</td>\n",
       "      <td>337</td>\n",
       "      <td>266</td>\n",
       "      <td>317</td>\n",
       "      <td>226</td>\n",
       "      <td>303</td>\n",
       "      <td>196</td>\n",
       "      <td>219</td>\n",
       "      <td>202</td>\n",
       "      <td>168</td>\n",
       "      <td>201</td>\n",
       "      <td>235</td>\n",
       "      <td>221</td>\n",
       "      <td>270</td>\n",
       "      <td>262</td>\n",
       "      <td>234</td>\n",
       "      <td>314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102466751</td>\n",
       "      <td>25</td>\n",
       "      <td>17</td>\n",
       "      <td>34</td>\n",
       "      <td>22</td>\n",
       "      <td>24</td>\n",
       "      <td>19</td>\n",
       "      <td>23</td>\n",
       "      <td>15</td>\n",
       "      <td>21</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>16</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107985730</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100302278</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      GeneID  GSM4996084  GSM4996085  GSM4996086  GSM4996087  GSM4996088  \\\n",
       "0  100287102           2           5           3           2           5   \n",
       "1     653635         244         236         337         266         317   \n",
       "2  102466751          25          17          34          22          24   \n",
       "3  107985730           1           1           1           0           0   \n",
       "4  100302278           0           0           0           0           0   \n",
       "\n",
       "   GSM4996089  GSM4996090  GSM4996091  GSM4996092  GSM4996093  GSM4996094  \\\n",
       "0           2           4           2           3           2           2   \n",
       "1         226         303         196         219         202         168   \n",
       "2          19          23          15          21          15          19   \n",
       "3           0           1           0           0           0           0   \n",
       "4           0           0           0           0           0           0   \n",
       "\n",
       "   GSM4996095  GSM4996096  GSM4996097  GSM4996098  GSM4996099  GSM4996100  \\\n",
       "0           4           3           2           2           3           3   \n",
       "1         201         235         221         270         262         234   \n",
       "2          22          15          17          16          18          17   \n",
       "3           0           0           0           0           0           0   \n",
       "4           0           0           0           0           0           0   \n",
       "\n",
       "   GSM4996101  \n",
       "0           3  \n",
       "1         314  \n",
       "2          23  \n",
       "3           1  \n",
       "4           0  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Load the bulk RNA-Seq data\n",
    "# Here we have the raw counts, which is the exact input we want for DESeq2.\n",
    "# Raw counts are generated by tools such as featureCounts or HTSeq-count.\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "counts_path = Path(\"..\") / \"data\" / \"GSE164073_raw_counts_GRCh38.p13_NCBI.tsv.gz\"\n",
    "counts = pd.read_csv(counts_path, sep=\"\\t\")  # pandas auto-detects .gz compression\n",
    "\n",
    "# Observe the data\n",
    "counts.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
